### PR TITLE
Changes for new properties model

### PR DESCRIPTION
**Pull request by vijaynalawade**
**What kind of change does this PR introduce?** (check one with "x")
```
[] Bugfix
[x] Feature
[] Code style update (formatting, local variables)
[] Refactoring (no functional changes, no api changes)
[] Other... Please describe:
```

**Fixes**: #
1. Handles new properties model as described at https://github.com/TIBCOSoftware/flogo/wiki/Application-Properties
2. Updated resolver to handle $property[VAR] and $env[VAR]
3. Added resolution logic for property value e.g. value now can be sourced from $env[VAR]
   `
   {
      "name": "PROP_BOOL",
      "type": "boolean",
      "value": "$env[BOOL_ENV_VAR]"
    }
   `

**What is the current behavior?**

**What is the new behavior?**
Following properties will be handled at runtime:
`"properties": [
    {
      "name": "PROP_BOOL",
      "type": "boolean",
      "value": true
    },
    {
      "name": "PROP_DOUBLE",
      "type": "double",
      "value": 2.5
    },
    {
      "name": "PROP_INT",
      "type": "integer",
      "value": 2
    },
    {
      "name": "PROP_LONG",
      "type": "long",
      "value": 2
    },
    {
      "name": "PROP_STRING",
      "type": "string",
      "value": "This is string"
    }
  ]
`

`
"properties": {
    "PROP_BOOL": true,
    "PROP_DOUBLE": 2.0,
    "PROP_INT": 2,
    "PROP_LONG": 2,
    "PROP_STRING": "This is string"
  }
`

### CODE REVIEW

File: core/data/resolve.go [hunk 1]
```diff
@@ -125,14 +125,23 @@ type ResolutionDetails struct {
...
 
-	dotIdx := strings.Index(toResolve, ".")
+	bracketIdx := strings.Index(toResolve, "]")
+	exprLen := len(toResolve)
+	if bracketIdx == (exprLen-1) && (strings.HasPrefix(toResolve, "property") || strings.HasPrefix(toResolve, "env")) {
+		//$property[] or $env[] resolution
+		itemIdx := strings.Index(toResolve, "[")
+		details.ResolverName = toResolve[:itemIdx]
+		details.Property = toResolve[itemIdx+1 : exprLen-1]
+		return details, nil
```
Comment (lixingwang) [inline core/data/resolve.go]: have issue to handle the array field case: `$property[propertyName].fieldname.array[0]`

File: app/config.go [hunk 2]
```diff
@@ -9,17 +9,18 @@ import (
 	"github.com/TIBCOSoftware/flogo-lib/core/action"
 	"github.com/TIBCOSoftware/flogo-lib/core/data"
 	"github.com/TIBCOSoftware/flogo-lib/core/trigger"
+	"github.com/pkg/errors"
```
Comment (fm-tibco) [inline app/config.go]: Should just use "errors" package from standard lib

File: app/config.go [hunk 3]
```diff
@@ -9,17 +9,18 @@ import (
...
-	Version     string                 `json:"version"`
-	Description string                 `json:"description"`
-	Properties  map[string]interface{} `json:"properties"`
-	Triggers    []*trigger.Config      `json:"triggers"`
-	Resources   []*resource.Config     `json:"resources"`
+	Name        string             `json:"name"`
+	Type        string             `json:"type"`
+	Version     string             `json:"version"`
+	Description string             `json:"description"`
+	Properties  interface{}        `json:"properties"`
```
Comment (fm-tibco) [inline app/config.go]: not sure if anyone actually used the previous property support, wasn't really advertised.  Maybe we should just make it []Attribute, @mellistibco thoughts?

Comment (mellistibco) [reply app/config.go]: I think its ok to just discard the previous property support, I suspect very few people were using it...

File: core/data/resolve.go (hunk 1 above)
Comment (vijaynalawade) [reply core/data/resolve.go]: @lixingwang : Currently, we support only primitive types. As use cases evolve, we will add support for Object and Array in future.

File: app/config.go (hunk 3 above)
Comment (vijaynalawade) [reply app/config.go]: @mellistibco @fm-tibco : This is fixed in now. Please review updated PR.

File: app/config.go [hunk 4]
```diff
@@ -18,14 +17,20 @@ type Config struct {
...
-	Properties  interface{}        `json:"properties"`
+	Properties  []AppProperty      `json:"properties"`
 	Triggers    []*trigger.Config  `json:"triggers"`
 	Resources   []*resource.Config `json:"resources"`
 
 	//for backwards compatibility
 	Actions []*action.Config `json:"actions"`
 }
 
+type AppProperty struct {
```
Comment (fm-tibco) [inline app/config.go]: Why not use Attribute?  It has the same exact members

Comment (vijaynalawade) [reply app/config.go]: @fm-tibco : Make sense. Check updated PR.

**Pull request merged**